fix(@embark/core): Fix `—template` URL support

emizzle opened the pull request:
Support for all template URL syntax as spec’d in https://our.status.im/embark-3-2-much-more-than-a-moonshot/ (Everyone Gets A Template section).

Unit tests were added for all missing cases.

# Test results
| Test case        | Command           | Successful?  | Output |
| :----------------------------------------- | :--------------------------------------------------------- | :---: | :-------- |
| ~~1a. Normal React template, no URL~~  			| ~~`embark new 1-normal-template-react --template react`~~ | ~~No~~ | ~~see below~~ |
```
Installing template from https://github.com/embark-framework/embark-react-template/tree/4.0
Downloading template...
Download failed, response code 404
Retrying with the default branch...
It may not be compatible with your Embark version 4.0.0-alpha.2
Installing template from https://github.com/embark-framework/embark-react-template
Downloading template...
Download failed, response code 404
Does the template really exist?
Embark's supported templates: https://embark.status.im/templates/
```
~~**Failure reason**: https://github.com/embark-framework/embark-react-template repo doesn't exist~~
~~**Do we know what happened to this repo?**~~

** **UPDATE** **
This was never really meant to be a test case :) The react template can be installed using the command:
```
embark demo
```

| Test case        | Command           | Successful?  | Output |
| :----------------------------------------- | :--------------------------------------------------------- | :---: | :-------- |
| 1b. Normal TypeScript template, no URL  			| `embark new 1-normal-template-typescript --template typescript` |  Yes | see below |
```
Installing template from https://github.com/embark-framework/embark-typescript-template/tree/4.0
Downloading template...
Download failed, response code 404
Retrying with the default branch...
It may not be compatible with your Embark version 4.0.0-alpha.2
Installing template from https://github.com/embark-framework/embark-typescript-template
Downloading template...
Installing packages...
Init complete

App ready at 1-normal-template-typescript
```
| Test case        | Command           | Successful?  | Output |
| :----------------------------------------- | :--------------------------------------------------------- | :---: | :-------- |
| 1c. Normal Vyper template, no URL  			| `embark new 1-normal-template-vyper --template vyper` | Yes |see below |
```
Installing template from https://github.com/embark-framework/embark-vyper-template/tree/4.0
Downloading template...
Download failed, response code 404
Retrying with the default branch...
It may not be compatible with your Embark version 4.0.0-alpha.2
Installing template from https://github.com/embark-framework/embark-vyper-template
Downloading template...
Installing packages...
Init complete

App ready at 1-normal-template-vyper
```
| Test case        | Command           | Successful?  | Output |
| :----------------------------------------- | :--------------------------------------------------------- | :---: | :-------- |
| 1d. Normal Vue.js template, no URL  			| `embark new 1-normal-template-vue --template vue` |  Yes |see below|
```
Installing template from https://github.com/embark-framework/embark-vue-template/tree/4.0
Downloading template...
Download failed, response code 404
Retrying with the default branch...
It may not be compatible with your Embark version 4.0.0-alpha.2
Installing template from https://github.com/embark-framework/embark-vue-template
Downloading template...
Installing packages...
Init complete

App ready at 1-normal-template-vue
```
| Test case        | Command           | Successful?  | Output |
| :----------------------------------------- | :--------------------------------------------------------- | :---: | :-------- |
| 1e. Normal Bamboo template, no URL  			| `embark new 1-normal-template-bamboo --template bamboo` |  Yes |see below|
```
Installing template from https://github.com/embark-framework/embark-bamboo-template/tree/4.0
Downloading template...
Download failed, response code 404
Retrying with the default branch...
It may not be compatible with your Embark version 4.0.0-alpha.2
Installing template from https://github.com/embark-framework/embark-bamboo-template
Downloading template...
Installing packages...
Init complete

App ready at 1-normal-template-bamboo
```
| Test case        | Command           | Successful?  | Output |
| :----------------------------------------- | :--------------------------------------------------------- | :---: | :-------- |
| 2. Github repo download  	 			| `embark new 2-github-repo --template status-im/dreddit-devcon`      |   Yes  |see below|
```
Installing template from https://github.com/status-im/dreddit-devcon
Downloading template...
Installing packages...
Init complete

App ready at 2-github-repo
```
| Test case        | Command           | Successful?  | Output |
| :----------------------------------------- | :--------------------------------------------------------- | :---: | :-------- |
| 3a. Github template with non-existant branch  	| `embark new 3-github-branch --template status-im/dreddit-devcon#start-here`      |  Yes   |
```
Installing template from https://github.com/status-im/dreddit-devcon/tree/start-here
Downloading template...
Download failed, response code 404
Does the template really exist?
```
| Test case        | Command           | Successful?  | Output |
| :----------------------------------------- | :--------------------------------------------------------- | :---: | :-------- |
| 3b. Github template with other branch  	| `embark new 3-github-branch --template status-im/dreddit-devcon#start`      |  Yes  |see below|
```
Installing template from https://github.com/status-im/dreddit-devcon/tree/start
Downloading template...
Installing packages...
Init complete

App ready at 3-github-branch
```
| Test case        | Command           | Successful?  | Output |
| :----------------------------------------- | :--------------------------------------------------------- | :---: | :-------- |
| 4. Github download with git url  		| `embark new 4-github-git-url --template git@github.com/status-im/dreddit-devcon`      |  No  |see below|
```
Unsupported template name or git host URL
```
**Failure reason:** Incorrect syntax. Repo owner need to postfix the host, ie `git@github.com:status-im/dreddit-devcon`. Fixed in this PR.

| Test case        | Command           | Successful?  | Output |
| :----------------------------------------- | :--------------------------------------------------------- | :---: | :-------- |
| 5. Github download with https url  		| `embark new 5-github-https-url --template https://github.com/status-im/dappcon-workshop-dapp` |  Yes  |see below|
```
Installing template from https://github.com/status-im/dappcon-workshop-dapp
Downloading template...
Installing packages...
Init complete

App ready at 5-github-https-url
```
| Test case        | Command           | Successful?  | Output |
| :----------------------------------------- | :--------------------------------------------------------- | :---: | :-------- |
| 6. Github download with no protocol  	| `embark new 6-github-url-no-protocol --template github.com/status-im/dappcon-workshop-dapp` |  No  |see below|
```
Unsupported template name or git host URL
```
**Failure reason:** Incorrect syntax. Can remove the github host, ie `status-im/dappcon-workshop-dapp`. Fixed in this PR.

| Test case        | Command           | Successful?  | Output |
| :----------------------------------------- | :--------------------------------------------------------- | :---: | :-------- |
| 7. Github repo with no branch, and a default branch that is not `master` | `embark new 7-github-default-branch-not-master --template status-im/dreddit-devcon`      |  Yes  |see below|
```
Installing template from https://github.com/status-im/dreddit-devcon
Downloading template...
Installing packages...
Init complete

App ready at 7-github-default-branch-not-master
```

alaibe commented on src/lib/utils/template_generator.js:
```diff
@@ -128,6 +128,18 @@ class TemplateGenerator {
+      'git@github.com/': 'git@github.com:',
+      'git@bitbucket.org/': 'git@bitbucket.org:',
+      'git@gitlab.com/': 'git@gitlab.com:',
+      'github.com/': '',
```
what happens in that case?
We replace with nothing?

emizzle replied:
In that case, the `hosted-git-info` library picks it up as a github shortcut and is then able to parse it properly.

michaelsbradleyjr commented:
Personally I’d rather us update the docs (if they’re not already), and/or add a note to that old announcement to the effect that we support what hosted-git-info supports and not the extra formats. That’s less maintenance for us in the long term, and it’s all the formats yarn and npm support because they rely on hosted-got-info too (it’s where I got the idea).

I don’t mean to knock this good work, though.

PascalPrecht commented on src/lib/utils/template_generator.js:
```diff
@@ -128,6 +128,18 @@ class TemplateGenerator {
   getExternalProject(uri) {
     let url, folder, hgi;
     let fallback, url_fallback, folder_fallback, hgi_fallback, embarkVersion;
+    const replacements = {
```
Only thing I'd suggest here would be to move the `replacements` out of `getExternalProject()` because everytime this function is call, we re-initialize those, which isn't needed because they are static.

PascalPrecht commented on src/lib/utils/template_generator.js:
```diff
@@ -128,6 +128,18 @@ class TemplateGenerator {
+      'git@gitlab.com/': 'git@gitlab.com:',
+      'github.com/': '',
+      'bitbucket.org/': '',
+      'gitlab.com/': ''
```
> In that case, the hosted-git-info library picks it up as a github shortcut and is then able to parse it properly.

I think it'd be useful to have this (or a shorter version) comment appended to one of these last two replacements.

Happy to trade it against the

> // reformat uri before parsing with hosted-git-info. Allows for further syntax support.

Comment a few lines below :)

michaelsbradleyjr commented:
@PascalPrecht I introduced the switch from the *ad hoc* url format (introduced by the bounty developer) to `hosted-git-info` before we published `3.2` to the npm registry. But, the announcement blog post had already been finalized and wasn't updated prior the package being published.  In other words, our published software never actually worked the way the blog post said it did.  You can verify with:
```
npm pack embark@3.2.1 && tar xvzf embark-3.2.1.tgz && less package/lib/utils/template_generator.js
```
We published a `3.2.0` but immediately deprecated it as the tarball contained a lot of unnecessary files. After updating `.npmignore` we published `3.2.1` a few minutes later.

There was a discussion in Slack (and it would be in the archives) at the time the blog post went live about how it had some outdated info re: template URLs, but I think we shrugged it off at the time. I had made a point to update the docs with examples based on what `hosted-git-info` actually supports.

I think the bounty developer was trying to imitate `hosted-git-info` — he was probably familiar with using URLs with `npm install` and `package.json`. However, he didn't get it quite right. I realized that during QA for 3.2 and that's when I did some research, learned about `hosted-git-info` by looking at npm's source code, and then refactored using that package.

PascalPrecht commented:
Ah good to know!

The fix itself is pretty small though, so I don't think it'll hurt to keep in in.

PascalPrecht commented:
> If we could instead have marketing change a couple of lines in the announcement post then it will reflect reality for all 3.2.x releases and v4.

Meaning that we'd update the post to not include these supported variations and drop this PR?

I'd personally vote to land this and ideally have it backported to 3.2.x. But I don't have super strong feelings about it otherwise.

emizzle commented:
Replacements has been moved to a constant, and I've changed them so that only `github.com/` is replaced with an empty string. The other URLs are replaced with their `hosted-git-info` equivalents. 

To support this, I have also added a couple more unit tests for `bitbucket.org/gitlab.com` shortcuts.

I agree that we could do all of the above:
1. eventually look to get rid of `--templates`
2. backport this to a 3.2.x release
3. Update the marketing blog post to remove the non-standard URL formats (but still support the non-standard formats in case anyone is using them)
